File: 03_Combinatorics_Search/30_Subset/base_code.py
# 1. 백트래킹을 이용한 정석 구현 (이진 트리 구조)
def generate_subsets(arr):
    result = []
    
    def backtrack(index, path):
        # 1. 종료 조건: 배열의 끝까지 모두 "넣을지 말지" 결정을 내렸다면 결과에 추가
        if index == len(arr):
            result.append(path[:])
            return
        
        # 2. 현재 원소를 '포함하는' 경우
        path.append(arr[index])
        backtrack(index + 1, path)
        path.pop() # 상태 복구
        
        # 3. 현재 원소를 '포함하지 않는' 경우
        backtrack(index + 1, path)
        
    backtrack(0, [])
    return result

# =====================================================================
# 2. 비트마스킹을 이용한 부분집합 (코딩 테스트 심화용)
# 원소의 포함 여부를 2진수의 0과 1로 표현하여 훨씬 빠르고 간결하게 구현합니다.
def generate_subsets_bitmask(arr):
    n = len(arr)
    result = []
    # 1 << n 은 2^n 을 의미합니다. (원소가 3개면 0부터 7까지)
    for i in range(1 << n):
        subset = []
        for j in range(n):
            # i의 j번째 비트가 1인지 확인 (1이면 해당 원소 포함)
            if i & (1 << j):
                subset.append(arr[j])
        result.append(subset)
    return result

# 사용 예시
if __name__ == "__main__":
    data = [1, 2, 3]
    
    print("백트래킹 부분집합:")
    for s in generate_subsets(data):
        print(s)
        
    print("\n비트마스킹 부분집합:")
    for s in generate_subsets_bitmask(data):
        print(s)

File: 03_Combinatorics_Search/30_Subset/BaseCode.java
import java.util.ArrayList;
import java.util.List;

public class BaseCode {

    /**
     * 부분집합을 구하는 메인 헬퍼 함수
     */
    public static void generateSubsets(int[] arr) {
        boolean[] visited = new boolean[arr.length]; // 각 원소의 포함/미포함 상태
        backtrack(arr, visited, 0);
    }

    /**
     * 백트래킹을 이용한 부분집합 재귀 함수
     * @param index 현재 포함 여부를 결정할 배열의 인덱스
     */
    private static void backtrack(int[] arr, boolean[] visited, int index) {
        // 1. 종료 조건: 모든 원소에 대해 포함/미포함 결정을 내렸을 때
        if (index == arr.length) {
            printSubset(arr, visited);
            return;
        }

        // 2. 현재 원소를 부분집합에 '포함'하는 경우
        visited[index] = true;
        backtrack(arr, visited, index + 1);

        // 3. 현재 원소를 부분집합에 '포함하지 않는' 경우
        visited[index] = false;
        backtrack(arr, visited, index + 1);
    }

    /**
     * visited 배열이 true인 원소만 골라서 출력하는 헬퍼 함수
     */
    private static void printSubset(int[] arr, boolean[] visited) {
        List<Integer> subset = new ArrayList<>();
        for (int i = 0; i < arr.length; i++) {
            if (visited[i]) {
                subset.add(arr[i]);
            }
        }
        System.out.println(subset);
    }

    // 사용 예시
    public static void main(String[] args) {
        int[] data = {1, 2, 3};
        
        System.out.println("자바 백트래킹을 이용한 부분집합:");
        generateSubsets(data);
    }
}